libraries/rush-lib/src/api/test/RushProjectSelector.test.ts

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { RushConfiguration } from '../RushConfiguration';
+import { RushConfigurationProject } from '../RushConfigurationProject';
+import { RushProjectSelector } from '../RushProjectSelector';
+
+function createProjectSelector(): RushProjectSelector {
+  const rushJsonFile: string = `${__dirname}/repo/rush-pnpm.json`;
+  const rushConfig: RushConfiguration = RushConfiguration.loadFromConfigurationFile(rushJsonFile);
+  const projectSelector: RushProjectSelector = new RushProjectSelector(rushConfig, {
+    gitSelectorParserOptions: {
+      includeExternalDependencies: true,
+      enableFiltering: false
+    }
+  });
```

Comment:
<img width="943" alt="Screenshot 2023-09-13 at 11 22 56 PM" src="https://github.com/microsoft/rushstack/assets/107583986/bbfa11f2-108f-45c3-ae86-390587699f71">

Reply:
What is a "selector"? Looking at these docs, a "selector" is an engine that chooses a list of Rush projects, kind of like how a CSS selector chooses DOM elements.

You have created a class called `RushProjectSelector` whose state is a **rush.json** configuration plus some options controlling how the Git selector behaves.  If we use SQL terminology, this class is like a SQL table plus some global options that affect query behavior.  It has the ability to execute queries, but the class is not a query -- instead the actual query is passed into `projectSelector.selectExpression()`.

Is a "selector" that query? Well, the query's type is `SelectorExpression` (not to be confused with `ExpressionSelector`, let's ignore that for now).

What is a "selector expression"? Well apparently it can be a selector:

```ts
const projects: ReadonlySet<RushConfigurationProject> = await projectSelector.selectExpression({
  '--to': 'project1'
});
```

Is it then the case that `--to X` and `X` are both selectors? Is everything a "selector"?  Is a "selector" just any AST node for this query language?

Reply:
Constructive suggestions:

1. Per my [other comment](https://github.com/microsoft/rushstack/pull/4271/files#r1324763400), let's return to the idea that a **"selector"** is a leaf node like `git:origin/main` or `tag:release`.
2. An expression like `to(tag:release and git:origin/main)` should get some other name, like **selection expression**.  `to()` itself is really a **"selection function"** but its JSON notation is an expression involving that function. 


`--to` is the CLI parameter representation of `to()`. I feel like the JSON design went off the rails a bit by putting `--to` in a JSON key.  JSON keys are JavaScript variable names. We don't write `let --to: string;`.  Of course we do want the JSON syntax to be concise to write, but it also needs to be an easy API to use.  The API scenarios are things like:
- a program that generates a selection expression 
- a program that optimizes a selection expression by reducing it to a more efficient form

Such programs aren't very easy to write if we can't even do:

```ts
switch (node.kind) {
  case 'union': . . .
  case 'intersect': . . .
  . . .
}
```

and instead must do:

```ts
    if (isUnion(expr)) {
      return this._evaluateUnion(expr, context);
    } else if (isIntersect(expr)) {
      return this._evaluateIntersect(expr, context);
    } else if (isSubtract(expr)) {
      return this._evaluateSubtract(expr, context);
    } else if (isParameter(expr)) {
      return this._evaluateParameter(expr, context);
    } else if (isDetailedSelector(expr)) {
```
3. So I recommend going back to a more conventional AST design.
4. Rename `RushProjectSelector.selectExpression` to something like `ProjectQueryContext.select()`